### PR TITLE
2030 5 agent

### LINKED ISSUE #3069: 2030.5 agent point creation
Configuration of points that will be read and updated from an actual device through the platform driver.

### PR DESCRIPTION
**Pull request by craig8**
# Description

This is a pull request to add the 2030.5 agent to the volttron repository.

Fixes #3069, #3070, #3071

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

This has been tested with the gridappsd-2030-5 server as of 02/22/2023.  It has been tested with points being generated and posted to the volttron platform via the simulated converter.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/core/IEEE_2030_5/ieee_2030_5/client.py [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+            href = getattr(my_response, "href")
+            outer_map[href] = my_response
+            for inner in getattr(my_response, list_prop):
+                href = getattr(inner, "href")
+                inner_map[href] = inner
+            
+        
+    def _get_device_capabilities(self, endpoint: str) -> m.DeviceCapability:
+        dcap: m.DeviceCapability = self.__get_request__(endpoint)
+        if self._response_status != 200:
```
Comment (schandrika) [inline services/core/IEEE_2030_5/ieee_2030_5/client.py:310]: some parts use _is_ok and and some check for response code. This particular place also assumes response is set in self._response_status. Better design might be for get/put methods to return response and _is_ok to accept that as param and return True/False

File: services/core/IEEE_2030_5/ieee_2030_5/agent.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        _log.debug(f"Dcap was updated by {sender}")
+
+    def configure(self, config_name, action, contents):
+        """
+        Called after the Agent has connected to the message bus. If a configuration exists at startup
+        this will be called before onstart.
+
+        Is called every time the configuration in the store changes.
+        """
+        config = self._default_config.copy()
```
Comment (schandrika) [inline services/core/IEEE_2030_5/ieee_2030_5/agent.py:139]: Configure method is not updating self._mirror_usage_point_list and default_config points to that.

File: services/core/IEEE_2030_5/ieee_2030_5/agent.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+                reading = self._mup_readings[reading_mRID]
+                for rs_index, rs in enumerate(reading.MirrorReadingSet):
+                    rs = reading.MirrorReadingSet[rs_index]
+                    rs.Reading.append(m.Reading(value=points.points[pt["subscription_point"]]))
+                    start = rs.timePeriod.start
+                    if start + self._mup_pollRate * 1000 > int(round(datetime.utcnow().timestamp())):
+                        self._times_published[reading_mRID] = self._times_published.get(reading_mRID, 0) + 1
+                        rs.mRID = "_".join([reading_mRID, str(self._times_published[reading_mRID])])
+
+                        new_reading_href = self._client.post_mirror_reading(reading)
```
Comment (schandrika) [inline services/core/IEEE_2030_5/ieee_2030_5/agent.py:227]: You have to post one reading at a time?

**Review by schandrika — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+[tool.yapfignore]
+ignore_patterns = [
+    ".env/**",
+    ".pytest_cache/**",
+    "dist/**",
+    "docs/**",
+]
+
+[tool.yapf]
```
[the remark this answers is not in the document]
Comment (craig8) [reply pyproject.toml:9]: We should have yapf installed on this repo's requirements as well and do a full reformat on all of the files.

File: services/core/IEEE_2030_5/ieee_2030_5/client.py [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+        Currently supports the following:
+        
+
+        
+        Args:
+
+            endpoint - /dcap by default but can be passed if there is a different entry point into hte
+                       system.
+        """
+        if not endpoint:
```
[the remark this answers is not in the document]
Comment (craig8) [reply services/core/IEEE_2030_5/ieee_2030_5/client.py:268]: The client should not connect until the "start" method is called

File: services/core/IEEE_2030_5/ieee_2030_5/agent.py (hunk 2 above)
Comment (craig8) [reply services/core/IEEE_2030_5/ieee_2030_5/agent.py:145]: On line 176 they are
self._mup_readings.clear()
self._mirror_usage_points.clear()

ed = self._client.end_device
self._mirror_usage_points.update(new_usage_points)
server_usage_points = self._client.mirror_usage_point_list()
self._mup_pollRate = server_usage_points.pollRate if server_usage_points.pollRate else self._mup_pollRate

File: services/core/IEEE_2030_5/ieee_2030_5/agent.py (hunk 3 above)
Comment (craig8) [reply services/core/IEEE_2030_5/ieee_2030_5/agent.py:228]: Nope....this has bee fixed

**Comment by craig8** [comment]
Follow on #3127

**Pull request closed without merging**